pkg/query/iter.go

```diff
@@ -89,87 +92,172 @@ type chunkSeries struct {
 		return chunks[i].MinTime < chunks[j].MinTime
 	})
-	l := *(*labels.Labels)(unsafe.Pointer(&lset)) // YOLO!
-
-	return &chunkSeries{lset: l, chunks: chunks, mint: mint, maxt: maxt}
+	return &chunkSeries{
+		lset:   *(*labels.Labels)(unsafe.Pointer(&lset)), // YOLO!
+		chunks: chunks,
+		mint:   mint,
+		maxt:   maxt,
+		aggr:   aggr,
+	}
 }
 
 func (s *chunkSeries) Labels() labels.Labels {
 	return s.lset
 }
 
 func (s *chunkSeries) Iterator() storage.SeriesIterator {
-	chks := make([]storepb.Chunk, 0, len(s.chunks))
-	for _, c := range s.chunks {
-		chks = append(chks, *c.Raw)
+	var sit storage.SeriesIterator
+	its := make([]chunkenc.Iterator, 0, len(s.chunks))
+
+	switch s.aggr {
+	case resAggrCount:
+		for _, c := range s.chunks {
+			its = append(its, getFirstIterator(c.Count, c.Raw))
+		}
+		sit = newChunkSeriesIterator(its)
+	case resAggrSum:
+		for _, c := range s.chunks {
+			its = append(its, getFirstIterator(c.Sum, c.Raw))
+		}
+		sit = newChunkSeriesIterator(its)
+	case resAggrMin:
+		for _, c := range s.chunks {
+			its = append(its, getFirstIterator(c.Min, c.Raw))
+		}
+		sit = newChunkSeriesIterator(its)
+	case resAggrMax:
+		for _, c := range s.chunks {
+			its = append(its, getFirstIterator(c.Max, c.Raw))
+		}
+		sit = newChunkSeriesIterator(its)
+	case resAggrCounter:
+		for _, c := range s.chunks {
+			its = append(its, getFirstIterator(c.Counter, c.Raw))
+		}
+		sit = downsample.NewCounterSeriesIterator(its...)
+	case resAggrAvg:
+		for _, c := range s.chunks {
+			if c.Raw != nil {
+				its = append(its, getFirstIterator(c.Raw))
+			} else {
+				sum, cnt := getFirstIterator(c.Sum), getFirstIterator(c.Count)
+				its = append(its, downsample.NewAverageChunkIterator(cnt, sum))
+			}
+		}
+		sit = newChunkSeriesIterator(its)
+	default:
+		return errSeriesIterator{err: errors.Errorf("unexpected result aggreagte type %v", s.aggr)}
 	}
-	return newChunkSeriesIterator(chks, s.mint, s.maxt)
+	return newBoundedSeriesIterator(sit, s.mint, s.maxt)
 }
 
-type nopSeriesIterator struct{}
-
-func (nopSeriesIterator) Seek(int64) bool      { return false }
-func (nopSeriesIterator) Next() bool           { return false }
-func (nopSeriesIterator) At() (int64, float64) { return 0, 0 }
-func (nopSeriesIterator) Err() error           { return nil }
+func getFirstIterator(cs ...*storepb.Chunk) chunkenc.Iterator {
+	for _, c := range cs {
+		if c == nil {
+			continue
+		}
+		chk, err := chunkenc.FromData(chunkEncoding(c.Type), c.Data)
+		if err != nil {
+			return errSeriesIterator{err}
+		}
+		return chk.Iterator()
+	}
+	return errSeriesIterator{errors.New("no valid chunk found")}
+}
 
-// chunkSeriesIterator implements a series iterator on top
-// of a list of time-sorted, non-overlapping chunks.
-type chunkSeriesIterator struct {
-	chunks     []storepb.Chunk
-	maxt, mint int64
+func chunkEncoding(e storepb.Chunk_Encoding) chunkenc.Encoding {
+	switch e {
+	case storepb.Chunk_XOR:
+		return chunkenc.EncXOR
+	}
+	return 255 // invalid
```

Comment:
why we cannnot have proto enum value fo this? `INVALID`?

Reply:
Also, maybe worth to propagate error here instead of special encoding type? Imagine in 2 months we will see 255/`invalid` encoding in some chunks? There is no way we can figure out whet in happen, etc... I would rather just fail fast (:

Reply:
A dedicated enum number in the proto would only help if the server explicitly set it to this value. This would generally not happen though as it should never return invalid chunks to begin with.

We just return this number for anything that's not recognized within our code here. Which can be any input value that's not `storepb.Chunk_XOR`.